doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -1581,33 +1606,88 @@ using the keyword :cmd:`Qed`.
 Attributes
 -----------
 
-Any vernacular command can be decorated with a list of attributes, enclosed
-between ``#[`` (hash and opening square bracket) and ``]`` (closing square bracket)
-and separated by commas ``,``. Multiple space-separated blocks may be provided.
+.. insertprodn all_attrs legacy_attrs
+
+.. prodn::
+   all_attrs ::= {* #[ {*, @attr } ] } {? @legacy_attrs }
+   attr ::= @ident {? @attr_value }
+   attr_value ::= = @string
+   | ( {*, @attr } )
+   legacy_attrs ::= {? {| Local | Global } } {? {| Polymorphic | Monomorphic } } {? Program } {? {| Cumulative | NonCumulative } } {? Private }
+
+Attributes modify the behavior of a command or tactic.
+Syntactically, most commands and tactics can be decorated with attributes, but
+attributes not supported by the command or tactic will be flagged as errors.
+
+The order of top-level attributes doesn't affect their meaning.  ``#[foo,bar]``, ``#[bar,foo]``,
+``#[foo]#[bar]`` and ``#[bar]#[foo]`` are equivalent.
 
-Each attribute has a name (an identifier) and may have a value.
-A value is either a :token:`string` (in which case it is specified with an equal ``=`` sign),
-or a list of attributes, enclosed within brackets.
+The legacy attributes (:n:`@legacy_attrs`) provide an older, alternate syntax
+for certain attributes.
 
 Some attributes are specific to a command, and so are described with
 that command. Currently, the following attributes are recognized by a
 variety of commands:
 
-``universes(monomorphic)``, ``universes(polymorphic)``
-    Equivalent to the ``Monomorphic`` and
-    ``Polymorphic`` flags (see :ref:`polymorphicuniverses`).
+.. attr:: universes(monomorphic)
+   :name: universes(monomorphic)
+
+   See :ref:`polymorphicuniverses`.
+
+.. attr:: Monomorphic
+
+   Legacy attribute equivalent to :attr:`universes(monomorphic)`.
+
+.. attr:: universes(polymorphic)
+   :name: universes(polymorphic)
+
+   See :ref:`polymorphicuniverses`.
+
+.. attr:: Polymorphic
+
+   Legacy attribute equivalent to :attr:`universes(polymorphic)`.
 
-``program``
-    Takes no value, equivalent to the ``Program`` flag
-    (see :ref:`programs`).
+.. attr:: program
 
-``global``, ``local``
-    Take no value, equivalent to the ``Global`` and ``Local`` flags
-    (see :ref:`controlling-locality-of-commands`).
+   See :ref:`programs`.
 
-``deprecated``
-    Takes as value the optional attributes ``since`` and ``note``;
-    both have a string value.
+.. attr:: Program
+
+   Legacy attribute equivalent to :attr:`program`.
+
+.. attr:: global
+
+   See :ref:`controlling-locality-of-commands`.
+
+.. attr:: Global
+
+   Legacy attribute equivalent to :attr:`global`.
+
+.. attr:: local
+
+   See :ref:`controlling-locality-of-commands`.
+
+.. attr:: Local
+
+   Legacy attribute equivalent to :attr:`local`.
+
+.. attr:: Cumulative
+
+   Legacy attribute THAT DOES WHAT?  Allowed only in a polymorphic context.
```

Comment:
See https://coq.gitlab.io/-/coq/-/jobs/442127533/artifacts/_install_ci/share/doc/coq/sphinx/html/addendum/universe-polymorphism.html#coq:cmd.cumulative. I am quite surprised that there is no `universes(cumulative)` syntax for this.

Note that `Cumulative`, `NonCumulative` and `Private` do not satisfy the above description of being "recognized by a variety of commands" (unless we consider that `Inductive` / `Record` / `Structure` / `Variant` represent a variety of commands, in which case this list is missing out on `universes(template)` and `universes(notemplate)`).

Reply:
I don't understand universe polymorphism or cumulativity well enough to write up for the manual.  I don't see where they are defined with any clarity in the manual.  The `Cumulative` command is only described as "Declares the inductive as cumulative".  Indeed, I only have a vague understanding of universes.

> Note that `Cumulative`, `NonCumulative` and `Private` do not satisfy the above description of being "recognized by a variety of commands"

True.  My inclination has been to deem any flag that appears before the main command name as an attribute.  That has its plusses and minuses for readability (it's in the way if we show {? Cumulative | NonCumulative } in the main syntax, but it's somewhat buried if the reader has to look at the following text to see that the command does indeed take certain attributes).  I could go either way--though we already made a decision on this and it would be some wasted motion to change it.

Reply:
> I don't see where they are defined with any clarity in the manual

The section @Zimmi48 linked to seems pretty clear to me.  Note that there is a large block of text with a couple of examples describing cumulativity; if you keep reading up to the heading "Cumulativity Weak Constraints", that is all still about cumulativity.  In any case, the description of NonCumulative is quite easy: it says that for two instances of the same inductive type (family) to be convertible, all the universes must be equal.  Cumulative relaxes this constraint according to the variance of the universes.

Reply:
I can use what you said for NonCumulative.

For Cumulative, variance is "invariant, covariant or irrelevant".  How are each of these treated for conversion (in the simplest language possible)?  I would rather not guess.

Reply:
I'm surprised there is no contravariant.
Irrelevant means that conversation of inductive type (families) imposes no constraint on this universe.  See also the list example in the docs.
I believe invariant imposes an equality constraint.  I expect that `Inductive foo@{i} := bar (_ : Type{i} -> Type@{i})` would make `i` invariant, and then equating `foo@{i}` and `foo@{j}` judgmentally would equate `i` and `j`.  You should perhaps check this, and maybe the docs should be updated with this example (or a less contrived one).
Covariant will clearly impose a `<=` constraint, but directionally is more subtle.  Let me check the docs.

Reply:
Btw, "Cic" seems like the wrong capitalization in https://coq.gitlab.io/-/coq/-/jobs/442127533/artifacts/_install_ci/share/doc/coq/sphinx/html/language/cic.html#subtyping-rules , maybe we should open an issue about it?

Reply:
As far as I can tell, the section on conversation rules has not been updated to account for cumulative universes.  Here is my guess for covariance, which should be tested:
In addition to getting the conversion problem that am inductive family must be equal at two different universe instantiations, you can also get the conversation problem that a term in the inductive family at one universe instance must also live in the inductive at another instance.  Consider, for example, the conversation problem arising from `fun x : T@{i} => (x : T@{j})` for some inductive `T`.  If the universe in `T` is irrelevant, then this imposes no constraint, as in the funext example involving `eq`.  If the universe is invariant, then this imposes the constraint `i = j`.  If, however, the universe is covariant, that this imposes the constraint `i <= j`.  The prototypical example of an inductive with a covariant universe is `Record T@{i} := { ty : Type@{i} }.` A sightly more interesting example is `Record Dyn@{i} := { typ : Type@{i} ; val : typ }`.  Note that the universe argument to `Type` is covariant.

This is all me making educated guesses; you (or someone) should check these examples before taking my word for it, and/or get one of the devs responsible for universes ( @mattam82 or @skyskimmer or @amintimany , I believe) to confirm.

Also, if I'm correct, these examples should be added to the chapter on cumulatively

Reply:
> Btw, "Cic" seems like the wrong capitalization in ...., maybe we should open an issue about it?

Yeah, since CIC is an acronym it would normally be in all caps.